services/printing/pdf_to_bitmap_converter.cc

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2024 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/services/printing/pdf_to_bitmap_converter.h"
+
+#include <utility>
+#include <vector>
+
+#include "base/logging.h"
+#include "base/memory/shared_memory_mapping.h"
+#include "pdf/pdf.h"
+#include "third_party/skia/include/core/SkBitmap.h"
+#include "ui/gfx/geometry/size_conversions.h"
+
+namespace printing {
+
+PdfToBitmapConverter::PdfToBitmapConverter() = default;
+
+PdfToBitmapConverter::~PdfToBitmapConverter() = default;
+
+void PdfToBitmapConverter::GetBitmap(
+    base::ReadOnlySharedMemoryRegion pdf_region,
+    GetBitmapCallback callback) {
+  // Decode memory region as PDF bytes.
+  base::ReadOnlySharedMemoryMapping pdf_map = pdf_region.Map();
+  if (!pdf_map.IsValid()) {
+    DLOG(ERROR) << "Failed to decode memory map for PDF thumbnail";
+    std::move(callback).Run(std::nullopt);
+    return;
+  }
+  auto pdf_buffer = pdf_map.GetMemoryAsSpan<const uint8_t>();
+
+  int page_count;
+  if (!chrome_pdf::GetPDFDocInfo(pdf_buffer, &page_count, nullptr)) {
+    DLOG(ERROR) << "Failed to get PDF document info";
+    std::move(callback).Run(std::nullopt);
+    return;
+  }
+
+  std::vector<SkBitmap> bitmaps;
+  for (int page_index = 0; page_index < page_count; page_index++) {
+    std::optional<gfx::SizeF> page_size =
+        chrome_pdf::GetPDFPageSizeByIndex(pdf_buffer, page_index);
+    if (!page_size.has_value()) {
+      DLOG(ERROR) << "Failed to get PDF page size";
+      std::move(callback).Run(std::nullopt);
+      return;
+    }
+    gfx::Size size = gfx::ToCeiledSize(*page_size);
+
+    SkBitmap bitmap;
+    const SkImageInfo info =
+        SkImageInfo::Make(size.width(), size.height(), kBGRA_8888_SkColorType,
+                          kOpaque_SkAlphaType);
+    if (!bitmap.tryAllocPixels(info, info.minRowBytes())) {
```

Comment:
what's the actual memory footprint of these bitmaps? A single A4 page @ 300dpi might take ~30MB in 8-bit RGBA. I'm afraid right now this will eat a lot of RAM, just think of 100+ page documents.

Please consider a queued conversion here, you're calling the text-recognition API on per-page basis anyway.

Reply:
Addressed in https://github.com/brave/brave-core/pull/22612/commits/cc93d167a0fef06502f306a041259f15c633be33

Reply:
Just did a profiling on a nine page doc https://docs.google.com/document/d/1OhFiLobEhBthcoaEbPIe_oNNmIq19lCYIuA_YEnmJAQ
<img width="555" alt="Screenshot 2024-03-19 at 17 12 42" src="https://github.com/brave/brave-core/assets/11330831/8c9e2ce4-fc5d-425b-83b4-8bee4c2aa8a1">

Reply:
recorded a new trace with legacy UI
<img width="863" alt="Screenshot 2024-03-19 at 17 29 41" src="https://github.com/brave/brave-core/assets/11330831/ef98ca85-f7b0-40d1-9e37-be47f3e4f4aa">

Reply:
> recorded a new trace with legacy UI

1. are you sure these are legit numbers? the exact moment memory metrics are acquired may miss the actual allocations.
2. what `chrome_pdf::GetPDFPageSizeByIndex` actually returns? Is it inches or something else? Have you looked at the resulting bitmaps you generate? I see a slightly different code in `PDFPageBitmap::Render` that scales the result of `GetPDFPageSizeByIndex` which in turn uses a different bitmap size than you do.

pls add some logs and look into the actual bitmap size, allocated memory and the generated bitmap to ensure everything is the way you expect it.

Also I still strongly believe that you should not generate all bitmaps upfront. Some logic to generate those bitmaps one by one when they are needed is the right thing to do. You shouldn't waste users RAM if you can avoid it. It's a performance concern, because you can swap-out background apps or worse - opened tabs.

Reply:
>1. are you sure these are legit numbers? the exact moment memory metrics are acquired may miss the actual allocations.

yep you are right, those profilings I tried are unstable and not always observable. I logged `computeByteSize()` for each bitmap for page size (420x595.92=>420x596) and it is 0.95MB per page and 9 pages makes it around 8.6MB.

>2. what chrome_pdf::GetPDFPageSizeByIndex actually returns? Is it inches or something else? Have you looked at the resulting bitmaps you generate? I see a slightly different code in PDFPageBitmap::Render that scales the result of GetPDFPageSizeByIndex which in turn uses a different bitmap size than you do.

It returns size 420x595.92 and ceiling to 420x596 in points
`PDFPageBitmap::Render` scales up its point size is to conform to CSS's 1pt unit so it can be used in headless testing.
We only need to render the page into image as it is.

>Also I still strongly believe that you should not generate all bitmaps upfront. Some logic to generate those bitmaps one by one when they are needed is the right thing to do. You shouldn't waste users RAM if you can avoid it. It's a performance concern, because you can swap-out background apps or worse - opened tabs.

We don't waste users resources in this use case because we have consistent upper limits between OCR logic and bitmap converter so each generated images are all actually used. If a source doc has more than 20 pages, the maximum bitmap allocation will be capped at 19MB because we only generate 20 images and do OCR on all of them.
I will still limit generating one bitmap per page when needed just in case this API will be used by other components which might not be aware of potential memory hogged if not being careful.

Reply:
> It returns size 420x595.92 and ceiling to 420x596 in points

So the bitmaps you're passing to OCR are exactly this size in pixels? Are you sure this is the right size to do the OCR properly? Did you check the quality of the OCR-generated text if the page is fully filled with text?

This all looks very wrong to me. The bitmap size is very small for OCR, it's basically a mm-to-pixel ratio of A2 page, i.e. 25.4 DPI (1 inch = 25.4mm).

![image](https://github.com/brave/brave-core/assets/5928869/7492dd3f-310f-41b3-b8c5-2232be156cb6)

Reply:
I have checked the text after OCR, and they are exactly same as what I see in original doc

Reply:
100% accuracy on https://docs.google.com/document/d/1OhFiLobEhBthcoaEbPIe_oNNmIq19lCYIuA_YEnmJAQ/edit

Reply:
The size(420x596) you see is original pdf size, image will be rendered in 300 dpi, if it is lower than 300 it will generate poor OCR result.

Reply:
>So the bitmaps you're passing to OCR are exactly this size in pixels? 
In points not pixels (1px = 0.75pt), so it is 560px x 794.66 px which is same size if you inspect the google doc canvas

Reply:
converting image and do OCR per page addressed in 4166f2aff4298bddddd8318efe168a83db39a5cb

Reply:
With 300 dpi, the dimension for rendering 560px x 795px image is 1.87in x 2.65in and the size `gfx::Size size = gfx::ToCeiledSize(*page_size);` (420pt x 596pt = 5.83in x 9.28in) passed into `chrome_pdf::RenderPDFPageToBitmap` should be sufficient enough to contain the image.

Reply:
okay, I think I figured what's going on and why things seems to be working fine, but I wasn't able to connect the dots.

1. `GetPDFPageSizeByIndex` returns the size in points, NOT pixels. Each point is 1/72 of an inch (see `chrome_pdf::CalculatePosition` and `printing::kPointsPerInch`).
2. You allocate the bitmap using points you get from this call, NOT pixels.
3. You call `chrome_pdf::RenderPDFPageToBitmap` with the bitmap you allocated and a requested dpi of 300, but the renderer can't really fit the 300 dpi image into the bitmap you passed.
4. The renderer recalculates width/height and renders the image at pixel-to-point dpi (72), ignoring 300 dpi you pass.

So after these manipulations you get 72 dpi image that happens to pass OCR. Will it work in documents with a smaller font size? Did you really want to run the OCR with 72 dpi images?

Either way, please add comments on what's actually going on and maybe increase the dpi to cover documents with smaller fonts.

Reply:
yep, we definitely need upscale to 300 dpi because when I changed font size from 11 to 5, the OCR result is totally wrong.

Reply:
Thanks for the summary, upscale bitmap to 300 dpi in https://github.com/brave/brave-core/pull/22612/commits/f942df0b2f5b9b25d09eddb998f3a11a4b526a74. And now the bitmap size for each page is 16.58MB which is the maximum bitmap allocation for the utility process because we do conversion and OCR on per page basis.